File: 21 Calendar/javascript/calendar.js
// CALENDAR
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//
function print(str)
{
	document.getElementById("output").appendChild(document.createTextNode(str));
}

function tab(space)
{
	var str = "";
	while (space-- > 0)
		str += " ";
	return str;
}

print(tab(32) + "CALENDAR\n");
print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
print("\n");
print("\n");
print("\n");

//       0, 31, 29  ON LEAP YEARS
var m = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

// VALUES FOR 1979 - SEE NOTES
for (i = 1; i <= 6; i++)
	print("\n");

d = -1;	// 1979 starts on Monday (0 = Sun, -1 = Monday, -2 = Tuesday)
s = 0;

for (n = 1; n <= 12; n++) {
	print("\n");
	print("\n");
	s = s + m[n - 1];
	str = "**" + s;
	while (str.length < 7)
		str += " ";
	for (i = 1; i <= 18; i++)
		str += "*";
	switch (n) {
		case  1:	str += " JANUARY "; break;
		case  2:	str += " FEBRUARY"; break;
		case  3:	str += "  MARCH  "; break;
		case  4:	str += "  APRIL  "; break;
		case  5:	str += "   MAY   "; break;
		case  6:	str += "   JUNE  "; break;
		case  7:	str += "   JULY  "; break;
		case  8:	str += "  AUGUST "; break;
		case  9:	str += "SEPTEMBER"; break;
		case 10:	str += " OCTOBER "; break;
		case 11:	str += " NOVEMBER"; break;
		case 12:	str += " DECEMBER"; break;
	}
	for (i = 1; i <= 18; i++)
		str += "*";
	str += (365 - s) + "**";
	     // 366 - s on leap years
	print(str + "\n");
	print("     S       M       T       W       T       F       S\n");
	print("\n");
	str = "";
	for (i = 1; i <= 59; i++)
		str += "*";    
	for (week = 1; week <= 6; week++) {
		print(str + "\n");
		str = "    ";
		for (g = 1; g <= 7; g++) {
			d++;
			d2 = d - s;
			if (d2 > m[n]) {
				week = 6;
				break;
			}
			if (d2 > 0)
				str += d2;
			while (str.length < 4 + 8 * g)
				str += " ";
		}
		if (d2 == m[n]) {
			d += g;
			break;
		}
	} 
	d -= g;
	print(str + "\n");
}

File: 21_Calendar/java/Calendar.java
/**
 * Game of Calendar
 * <p>
 * Based on the BASIC game of Calendar here
 * https://github.com/coding-horror/basic-computer-games/blob/main/21%20Calendar/calendar.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 * 
 * Converted from BASIC to Java by Darren Cardenas.
 */
 
public class Calendar {  

  private static final int NUM_WEEK_ROWS = 6;
  private static final int NUM_DAYS_PER_WEEK = 7;
  private static final int NUM_MONTHS_PER_YEAR = 12; 
  private static final int[] daysPerMonth = { 0, 31, 28, 31, 30, 31, 30, 
                                             31, 31, 30, 31, 30, 31 };  
    
  public void play() {
    
    showIntro();
    startGame();
    
  }  // End of method play  
    
  private static void showIntro() {
    
    System.out.println(" ".repeat(31) + "CALENDAR");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("\n\n");

  }  // End of method showIntro        
  
  private void startGame() {   

    int dayOfMonth = 0;
    int dayOfWeek = 0;
    int dayOfYear = 0;
    int daysTotal = 0;    
    int index = 0;
    int month = 0;
    int row = 0;    
    
    String lineContent = "";
    
    for (index = 1; index <= 6; index++) {     
      System.out.println("");      
    }
    
    daysTotal = -1;
    dayOfYear = 0;
    
    System.out.println("");
    
    // Begin loop through all months
    for (month = 1; month <= NUM_MONTHS_PER_YEAR; month++) {

      System.out.println("");
      
      dayOfYear = dayOfYear + daysPerMonth[month - 1];
      
      lineContent = String.format("** %-3d" + "*".repeat(18), dayOfYear);
      
      switch (month) {        
        case 1:
          lineContent += " JANUARY ";
          break;
        case 2:
          lineContent += " FEBRUARY";
          break;
        case 3:
          lineContent += "  MARCH  ";
          break;
        case 4:
          lineContent += "  APRIL  ";
          break;         
        case 5:          
          lineContent += "   MAY   ";
          break;         
        case 6:          
          lineContent += "   JUNE  ";
          break;         
        case 7:          
          lineContent += "   JULY  ";
          break;
        case 8:
          lineContent += "  AUGUST ";
          break;
        case 9:
          lineContent += "SEPTEMBER";
          break;
        case 10:
          lineContent += " OCTOBER ";
          break;
        case 11:
          lineContent += " NOVEMBER";
          break;
        case 12:
          lineContent += " DECEMBER";
          break;
        default:
          break;
      }
      
      lineContent += "*".repeat(18) + " " + (365 - dayOfYear) + "**";
      
      System.out.println(lineContent);
      System.out.println("");

      System.out.print("     S       M       T       W");
      System.out.println("       T       F       S");
      System.out.println("");
      
      System.out.println("*".repeat(59));

      // Begin loop through each week row
      for (row = 1; row <= NUM_WEEK_ROWS; row++) {
        
        System.out.println("");
        
        lineContent = "    ";
        
        // Begin loop through days of the week
        for (dayOfWeek = 1; dayOfWeek <= NUM_DAYS_PER_WEEK; dayOfWeek++) {
          
          daysTotal++;
          
          dayOfMonth = daysTotal - dayOfYear;
          
          if (dayOfMonth > daysPerMonth[month]) {
            row = 6;
            break;            
          }

          if (dayOfMonth > 0) {            
            lineContent += dayOfMonth;
          }
            
          while (lineContent.length() < (4 + 8 * dayOfWeek)) {            
            lineContent += " ";            
          }          
          
        }  // End loop through days of the week
        
        if (dayOfMonth == daysPerMonth[month]) {       
          row = 6;
          daysTotal += dayOfWeek;
          System.out.println(lineContent);
          break;
        }
        
        System.out.println(lineContent);
        
      }  // End loop through each week row
      
      daysTotal -= dayOfWeek;     
      
    }  // End loop through all months
    
    for (index = 1; index <= 6; index++) {     
      System.out.println("");      
    }    
    
  }  // End of method startGame     
  
  public static void main(String[] args) {
    
    Calendar game = new Calendar();
    game.play();
    
  }  // End of method main
    
}  // End of class Calendar
